### PR TITLE
Add support for Blake2s and Blake2b hash algorithms

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

Blake2 was initially proposed in [RFC 7693](https://datatracker.ietf.org/doc/html/rfc7693.html). It provides:
- hash and MAC algorithms;
- 256 or 512 bits of output;
- single or parallel execution.

PSA Crypto API added Blake2 into the specs in [version 1.5](https://arm-software.github.io/psa-api/crypto/1.5/):
- PSA_ALG_BLAKE2_MAC -> MAC algorithm
- PSA_ALG_BLAKE2S_HASH256 and PSA_ALG_BLAKE2B_HASH512 -> single execution
- PSA_ALG_BLAKE2SP_HASH256 and PSA_ALG_BLAKE2BP_HASH512 -> parallel execution

Adding support for all of this in a single PR would mean having a very large PR which is difficult to review. As a consequence the scope of this PR is limited to add only hash support (MAC is done in [this follow-up PR](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/898)) and for the single threaded case (i.e. no parallel).

Prerequisite PRs:
1. https://github.com/Mbed-TLS/mbedtls-framework/pull/316
2. https://github.com/Mbed-TLS/mbedtls/pull/10921

A follow-up PR (https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/898) will add support for MAC.

## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [ ] **changelog** provided
- [ ] **framework PR** provided: https://github.com/Mbed-TLS/mbedtls-framework/pull/316
- [ ] **TF-PSA-Crypto development PR** provided: this one
- [ ] **TF-PSA-Crypto 1.1 PR** not required because: it's an enhancement, so I would say this is not required
- [ ] **mbedtls development PR** provided: https://github.com/Mbed-TLS/mbedtls/pull/10921
- [ ] **mbedtls 4.1 PR** not required because: it's an enhancement, so I would say this is not required
- [ ] **mbedtls 3.6 PR** not required because: definitely no!
- **tests**  provided

### CODE REVIEW

File: include/psa/crypto_values.h [hunk 1]
```diff
@@ -947,6 +947,20 @@
...
+/** The BLAKE2b cryptographic hash with 512 bits of output (BLAKE2b-512). */
+#define PSA_ALG_BLAKE2B_HASH512                 ((psa_algorithm_t) 0x0200001E)
+
+/** Whether the specified algorithm is a BLAKE2 hash algorithm.
+ *
+ * \param alg An algorithm identifier: a value of type psa_algorithm_t.
+ *
+ * \return 1 if alg is a BLAKE2 hash algorithm, 0 otherwise.
+ */
+#define PSA_ALG_IS_BLAKE2_HASH(alg)     \
```
Comment (bjwtaylor) [inline include/psa/crypto_values.h:961]: I think this requires a parentheses around the expression as otherwise it is not safe to negate or compose

File: tests/suites/test_suite_md.function [hunk 2]
```diff
@@ -73,7 +78,8 @@ void md_to_from_psa()
     MD_PSA(MBEDTLS_MD_SHA3_224, PSA_ALG_SHA3_224);
     MD_PSA(MBEDTLS_MD_SHA3_256, PSA_ALG_SHA3_256);
     MD_PSA(MBEDTLS_MD_SHA3_384, PSA_ALG_SHA3_384);
-    MD_PSA(MBEDTLS_MD_SHA3_512, PSA_ALG_SHA3_512);
```
Comment (bjwtaylor) [inline tests/suites/test_suite_md.function:76]: Is it intentional to drop SHA3_512 support?

File: configs/crypto-config-symmetric-only.h [hunk 3]
```diff
@@ -22,6 +22,8 @@
 #define PSA_WANT_ALG_CHACHA20_POLY1305          1
 #define PSA_WANT_ALG_CMAC                       1
```
Comment (bjwtaylor) [inline configs/crypto-config-symmetric-only.h:23]: Putting this here as I can't comment on the commit messages commits 9d20167eda738062dba18702ba0173ef04f81f3d and 212901fe98d9b298eb0ea88b12a8a1db4685c959 refer blake2p should this be blake2b?

File: tests/suites/test_suite_blake2.b.data [hunk 4]
```diff
@@ -0,0 +1,770 @@
+Blake2b KAT #0 (msg len=0)
```
Comment (bjwtaylor) [inline tests/suites/test_suite_blake2.b.data:1]: It might be good to add the RFC 7693 Appendix A: BLAKE2b-512("abc") and  RFC 7693 Appendix B: BLAKE2s-256("abc") to exercise unkeyed mode.

File: configs/crypto-config-symmetric-only.h [hunk 5]
```diff
@@ -22,6 +22,8 @@
...
 #define PSA_WANT_ALG_CTR                        1
```
Comment (bjwtaylor) [inline configs/crypto-config-symmetric-only.h:24]: It will also probably need a changelog

File: drivers/builtin/include/tf-psa-crypto/private/blake2.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+typedef struct tf_psa_crypto_blake2s_context {
+    uint8_t MBEDTLS_PRIVATE(buf[64]);
+    uint32_t MBEDTLS_PRIVATE(state[8]);
+    uint32_t MBEDTLS_PRIVATE(processed_bytes)[2];
+    size_t MBEDTLS_PRIVATE(buf_idx);
+    size_t MBEDTLS_PRIVATE(outlen);
+} tf_psa_crypto_blake2s_context;
+
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_BLAKE2S_HASH256)
+
```
Comment (bjwtaylor) [inline drivers/builtin/include/tf-psa-crypto/private/blake2.h:38]: Is there a reason there is no doxygen for these functions? The whole implementation also lacks comments to explain it's operation and aid future maintainability. Is there a reason for this? I guess it's private, so we don't want the docs in the public documentation. It would still be useful to have some docs though. I've attached a patch of some docs i added to aid the review that might be a good starting point.

[documentation.patch](https://github.com/user-attachments/files/31173577/documentation.patch)

File: configs/crypto-config-symmetric-only.h (hunk 3 above)
Comment (bjwtaylor) [inline configs/crypto-config-symmetric-only.h:23]: Do the failing tests need investigating or is this expected?

File: tests/suites/test_suite_blake2.function [hunk 7]
```diff
@@ -0,0 +1,103 @@
+/* BEGIN_HEADER */
```
Comment (bjwtaylor) [inline tests/suites/test_suite_blake2.function:1]: It might be useful to extend the testing to inclulde:
  | Severity | Area | Missing coverage / action |
  | --- | --- | --- |
  | Major | Legacy MD API | Add expected-output tests for empty, block-1, block, and block+1 inputs; one-shot, multipart, clone divergence, HMAC/reset, and allocation failures. Cover both variants. |
  | Minor | PSA hash contracts | Add `psa_hash_compute()` tests, zero/short buffers, output sanitization, invalid expected hashes, invalid state transitions, abort/reuse, active-target cloning, and driver-error propagation. |
  | Minor | PSA HMAC | Add permanent BLAKE2s/BLAKE2b one-shot, multipart, verify, abort, and reset tests. Include malformed MACs, short buffers, policy/usage failures, and varied key lengths. |
  | Minor | PSA signatures | Add ECDSA message/pre-hash tests, RSA-PSS where supported, invalid signatures/digest lengths, short buffers, and policy/usage failures. Verify that RSA PKCS#1 v1.5 returns `PSA_ERROR_NOT_SUPPORTED`. |
  | Minor | Private BLAKE2 API | Test private one-shot functions, error propagation, shortened output lengths, varied key lengths, retry after a short finish buffer, and seeded counter carry. |
  | Contract decision | Invalid pointers | Decide whether `NULL` with a nonzero input/key length and a `NULL` output are caller preconditions. Document them or add validation and negative tests. |
  | Suggestion | Reusable test matrix | Instantiate the common PSA hash success and failure tests for every enabled fixed-output hash. |
  | Suggestion | Coverage documentation | Document that the exhaustive vectors validate digest correctness and boundaries, but not public API error contracts. |

File: drivers/builtin/src/blake2.c [hunk 8]
```diff
@@ -0,0 +1,382 @@
+/**
```
Comment (bjwtaylor) [inline drivers/builtin/src/blake2.c:1]: For the benefit other other reviewers this implementation flags as having similarities to nanonext 0.10.4 and PlantUML 1.2023.11 nanonext includes mbedtls and the simularities in PlantUML appear to be related to the fact they are both based RFC 7693 and not indicative of them being related implementations.

File: drivers/builtin/src/blake2.c [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+void tf_psa_crypto_blake2s_clone(tf_psa_crypto_blake2s_context *dst,
+                                 const tf_psa_crypto_blake2s_context *src)
+{
+    *dst = *src;
+}
+
+static void tf_psa_crypto_blake2s_compress(tf_psa_crypto_blake2s_context *ctx, int last)
+{
+    int i;
+    uint32_t v[16], m[16];
```
Comment (bjwtaylor) [inline drivers/builtin/src/blake2.c:112]: These two variables could contain sensitive data and although they are not available through the API, it probably makes sense to wipe them when your done with them

File: drivers/builtin/src/blake2.c [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+void tf_psa_crypto_blake2b_clone(tf_psa_crypto_blake2b_context *dst,
+                                 const tf_psa_crypto_blake2b_context *src)
+{
+    *dst = *src;
+}
+
+static void tf_psa_crypto_blake2b_compress(tf_psa_crypto_blake2b_context *ctx, int last)
+{
+    int i;
+    uint64_t v[16], m[16];
```
Comment (bjwtaylor) [inline drivers/builtin/src/blake2.c:286]: These two variables could contain sensitive data and although they are not available through the API, it probably makes sense to wipe them when your done with them

File: drivers/builtin/src/blake2.c [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+        B2B_G(2, 7,  8, 13, m[blake2_sigma[i][12]], m[blake2_sigma[i][13]]);
+        B2B_G(3, 4,  9, 14, m[blake2_sigma[i][14]], m[blake2_sigma[i][15]]);
+    }
+
+    for (i = 0; i < 8; ++i) {
+        ctx->state[i] ^= v[i] ^ v[i + 8];
+    }
+}
+
+void tf_psa_crypto_blake2b_update(tf_psa_crypto_blake2b_context *ctx,
```
Comment (bjwtaylor) [inline drivers/builtin/src/blake2.c:319]: Is there a reason we copy one byte at a time instead of copying one block at a time?

File: drivers/builtin/src/blake2.c [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+        B2S_G(2, 7,  8, 13, m[blake2_sigma[i][12]], m[blake2_sigma[i][13]]);
+        B2S_G(3, 4,  9, 14, m[blake2_sigma[i][14]], m[blake2_sigma[i][15]]);
+    }
+
+    for (i = 0; i < 8; ++i) {
+        ctx->state[i] ^= v[i] ^ v[i + 8];
+    }
+}
+
+void tf_psa_crypto_blake2s_update(tf_psa_crypto_blake2s_context *ctx,
```
Comment (bjwtaylor) [inline drivers/builtin/src/blake2.c:145]: Is there a reason we copy one byte at a time instead of copying one block at a time?

File: drivers/builtin/src/blake2.c (hunk 8 above)
Comment (bjwtaylor) [inline drivers/builtin/src/blake2.c:1]: It might be useful to put a scope at the top for future reference, something like 
 * This implements the sequential modes from RFC 7693, Sections 2.6 and 2.7.
 * Tree hashing parameters are not supported.
 Maybe also state that tree hashing is not supported due to it not being in the PSA API what do yoiu think?

**Review by bjwtaylor — COMMENTED** [review]
Some comments, questions and suggestions.

File: tests/suites/test_suite_psa_crypto_hash.blake2b.data [hunk 13]
```diff
@@ -0,0 +1,2051 @@
+PSA hash finish: Blake2b-512 KAT #0 (msg len=0)
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_psa_crypto_hash.blake2b.data]: Please provide the source of the test vectors.

File: tests/suites/test_suite_psa_crypto_hash.blake2s.data [hunk 14]
```diff
@@ -0,0 +1,2051 @@
+PSA hash finish: Blake2s-256 KAT #0 (msg len=0)
```
Comment (ronald-cron-arm) [inline tests/suites/test_suite_psa_crypto_hash.blake2s.data]: ditto

File: configs/crypto-config-symmetric-only.h (hunk 5 above)
Comment (valeriosetti) [reply configs/crypto-config-symmetric-only.h:24]: Yeah, that's marked as `TODO` in the description of the PR indeed :)

File: configs/crypto-config-symmetric-only.h (hunk 3 above)
Comment (valeriosetti) [reply configs/crypto-config-symmetric-only.h:23]: > Do the failing tests need investigating

Failing tests are due to missing changes in `mbedtls`. CI in https://github.com/Mbed-TLS/mbedtls/pull/10921 was green when [it was including this PR](https://github.com/Mbed-TLS/mbedtls/pull/10921#issuecomment-5292045443).

File: tests/suites/test_suite_md.function (hunk 2 above)
Comment (valeriosetti) [reply tests/suites/test_suite_md.function:76]: Oops, no, sorry

File: configs/crypto-config-symmetric-only.h (hunk 3 above)
Comment (valeriosetti) [reply configs/crypto-config-symmetric-only.h:23]: ouch, yes, sorry also for this one :/

File: drivers/builtin/include/tf-psa-crypto/private/blake2.h (hunk 6 above)
Comment (valeriosetti) [reply drivers/builtin/include/tf-psa-crypto/private/blake2.h:53]: > Is there a reason there is no doxygen for these functions? 

Lazyness during initial development followed by oversight when finalizing the work ;)

File: drivers/builtin/src/blake2.c (hunk 9 above)
Comment (valeriosetti) [reply drivers/builtin/src/blake2.c:107]: Not sure about this. `m` contains the input data that has been passed to `tf_psa_crypto_blake2s_update` so IMO that's not really secret material. As for `v` that's used to compute the new internal state that's stored in `ctx->state` so I'm not sure this is really leaking any information that's not available anyway in the context.
What exactly are `v` and `m` leaking?

I mean, the algorithm is deterministic, so ideally given the key (if any) and the input material anyone knows exactly what the output will look like. Isn't it?

File: drivers/builtin/src/blake2.c (hunk 10 above)
Comment (valeriosetti) [reply drivers/builtin/src/blake2.c:279]: Same as [above](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/867#discussion_r3905897541)

File: drivers/builtin/src/blake2.c (hunk 11 above)
Comment (valeriosetti) [reply drivers/builtin/src/blake2.c:312]: Honestly that was the design proposed in the RFC and I just took that as working solution. I knew it could perhaps be optimized, but I didn't bothered trying that at the time. I can do this now however!

> instead of copying one block at a time

On this please note that the input data is not necessarily provided in blocks whose size is multiple of 32 or 64, so some handling has to be done anyway.

File: drivers/builtin/src/blake2.c (hunk 12 above)
Comment (valeriosetti) [reply drivers/builtin/src/blake2.c:140]: Same as [above](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/867#discussion_r3906269484)

File: drivers/builtin/src/blake2.c (hunk 8 above)
Comment (valeriosetti) [reply drivers/builtin/src/blake2.c:1]: > Tree hashing parameters are not supported

IIRC that´s not part of the RFC 7963. It's something they implemented in the reference design in https://github.com/BLAKE2/BLAKE2, but it's not part of the RFC. See the end of[ this section](https://www.rfc-editor.org/info/rfc7693/#section-2.5).

File: tests/suites/test_suite_blake2.function (hunk 7 above)
Comment (valeriosetti) [reply tests/suites/test_suite_blake2.function:1]: - I added some MD testing. IMO this doesn't need to extensive because Blake2[s|b] already have their own testing. In MD we just need to ensure MD is able to call Blake2 when/if needed.
- No HMAC test added because HMAC for Blake2 is not supported yet (and it's not the same as other hash algorithms).
- Added some minimal testing for ECDSA, RSA-PSS and RSA PKCS#1 v1.5 using Blake2[s|p]

**Comment by valeriosetti** [comment]
Thanks to the merge of all prerequisite PRs now the CI is green and this PR is officially ready for review :)

File: drivers/builtin/src/blake2.c (hunk 11 above)
Comment (bjwtaylor) [reply drivers/builtin/src/blake2.c:312]: So, I think it's a nice to have. Not sure if it's essential though, I would see what the second reviewer thinks and at least add an additional issue that's it's further work.

File: drivers/builtin/src/blake2.c (hunk 8 above)
Comment (bjwtaylor) [reply drivers/builtin/src/blake2.c:1]: Maybe something like this then:
  /*
   * Implements sequential BLAKE2s and BLAKE2b as specified in RFC 7693.
   * Tree-hashing parameters and the parallel BLAKE2sp/BLAKE2bp variants
   * are not supported by this module.
   */
?

File: drivers/builtin/include/tf-psa-crypto/private/blake2.h [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+    size_t MBEDTLS_PRIVATE(buf_idx);               /*!< The number of bytes currently held in
+                                                      \c buf. */
+    size_t MBEDTLS_PRIVATE(outlen);                /*!< The configured digest length, in bytes. */
+} tf_psa_crypto_blake2s_context;
+
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_BLAKE2S_HASH256)
+
+/**
+ * \brief          This function initializes a BLAKE2s context and starts a
+ *                 new hash or HMAC operation.
```
Comment (bjwtaylor) [inline drivers/builtin/include/tf-psa-crypto/private/blake2.h:56]: Is this HMAC or native keyed BLAKE2? Ditto on 66:133:180:360

File: ChangeLog.d/blake2.txt [hunk 16]
```diff
@@ -0,0 +1,4 @@
+Features
+   * Support for Blake2 hash algorithms Blake2s and Blake2p has been added.
```
Comment (bjwtaylor) [inline ChangeLog.d/blake2.txt:2]: Should this be blake2b not 2p?

File: drivers/builtin/include/tf-psa-crypto/private/blake2.h (hunk 15 above)
Comment (valeriosetti) [reply drivers/builtin/include/tf-psa-crypto/private/blake2.h]: Oops, right, of course HMAC is not correct. OTOH using 'Blake2 MAC' all the times looks a bit too verbose, so I think I will go with just 'MAC'. Please, let me know if that doesn't look like a good idea to you.

File: ChangeLog.d/blake2.txt (hunk 16 above)
Comment (valeriosetti) [reply ChangeLog.d/blake2.txt]: Yes, sorry. I know the correct version is `2b`, but for some mysterious reason my mind keeps preferring `2p` :facepalm:

**Comment by valeriosetti** [comment]
I rebased to resolve a conflict on the `framework` reference. While at this, since current `framework` being referenced from `development` is also newer than what this PR needs, I dropped the commit updating the `framework` reference. This should reduce the potential conflict surface with other changes.
